probe-rs/src/vendor/raspberrypi/sequences/rp2040.rs

```diff
@@ -0,0 +1,135 @@
+//! Implement chip-specific dual-core reset for RP2040
+
+use std::sync::Arc;
+
+use crate::{
+    MemoryMappedRegister,
+    architecture::arm::{
+        ArmError,
+        armv6m::{Aircr, Demcr},
+        dp::{Ctrl, DpAddress, DpRegister},
+        memory::ArmMemoryInterface,
+        sequences::{ArmDebugSequence, cortex_m_wait_for_reset},
+    },
+};
+
+const SIO_CPUID_OFFSET: u64 = 0xd000_0000;
+const RESCUE_DP: DpAddress = DpAddress::Multidrop(0xf100_2927);
+
+// Performing a reset causes all cores to reset their SWD connections. The reset
+// takes place on core 0, however we still need to restore the core 1 protocol
+// characteristics such as overrun detection. To do this, we need to manually
+// specify the core 1 DP address.
+const CORE_1_DP: DpAddress = DpAddress::Multidrop(0x1100_2927);
+
+/// An empty struct that implements the default [ArmDebugSequence] methods
+/// to allow us to call default implementations from our derived function.
+#[derive(Debug)]
+struct DefaultArmDebugSequence;
+impl ArmDebugSequence for DefaultArmDebugSequence {}
+
+/// Debug implementation for RP2040
+#[derive(Debug)]
+pub struct Rp2040 {}
+
+impl Rp2040 {
+    /// Create a debug sequencer for a Raspberry Pi RP2040
+    pub fn create() -> Arc<Self> {
+        Arc::new(Rp2040 {})
+    }
+}
+
+impl ArmDebugSequence for Rp2040 {
+    fn debug_port_setup(
+        &self,
+        interface: &mut dyn crate::architecture::arm::communication_interface::DapProbe,
+        dp: DpAddress,
+    ) -> Result<(), ArmError> {
+        // Before we set up the requested DP, start and then stop the Default DP.
+        // This works around an issue where the multidrop DP becomes unavailable.
+        if DefaultArmDebugSequence
+            .debug_port_setup(interface, DpAddress::Default)
+            .is_err()
+        {
+            tracing::error!("Unable to connect to default address");
+        }
+        self.debug_port_stop(interface, DpAddress::Default)?;
+
+        // Delegate actual debug port setup to the default implementation.
+        DefaultArmDebugSequence.debug_port_setup(interface, dp)?;
+        Ok(())
+    }
+
+    fn reset_system(
+        &self,
+        core: &mut dyn ArmMemoryInterface,
+        core_type: crate::CoreType,
+        debug_base: Option<u64>,
+    ) -> Result<(), ArmError> {
+        tracing::trace!("Starting reset of RP2040");
+        // Only perform a system reset from core 0
+        let core_id = core.read_word_32(SIO_CPUID_OFFSET)?;
+        if core_id != 0 {
```

Comment:
I am not sure we should silently skip when we do this from the wrong core. Most users will not know what went wrong.

Reply:
Should it print an error? Or I suppose I could have it just poke AIRCR. This was causing issues in a previous version when I was using the watchdog timer rather than the recovery core, but perhaps it isn't an issue anymore.

Reply:
I mean, the question is whether this sequence has to be callable on core 1. If we do this internally, then we probably have to do it as it is right now. But in that case I would nudge the log to info. If we do not call this internally, I would prevent the user from being able to call it by returning an error.

Reply:
It is already solved :) The log was changed from debug to error. My message came after xobs by accident. I was okay with his answer.